feat: [OS-478] adds integrated VPLS and LSP sync functionality

haniotak opened the pull request:
### Description

lots of business logic to handle service sync, collecting NSO operations results, updating connection states
minor changes to nso.sync property
updates existing tests

### Checklist

- [x] PR Title format: `type: [OS-XXX] Short description` 
    - `type` is one of: 'build', 'ci', 'chore',  'docs',  'feat', 'fix',  'perf', 'refactor', 'revert', 'style', 'test' 
    - `OS-XXX` refers to a Jira issue. 
    - If this is a breaking change prefix the description with "BREAKING CHANGE:"
- [x] There is a related Jira issue for this pull request
- [x] Description should be a meaningful summary of the changes you are proposing
- [x] For breaking changes prefix the title with `"BREAKING CHANGE:"` and include details in the description
- [x] This PR includes tests related to these changes, or existing tests provide coverage
- [x] This PR has updated documentation as appropriate
- [x] `mvn clean package` runs successfully

### Notes
- Once approved, use the **squash and merge** option.
- Refer to the [development notes](https://github.com/esnet/oscars/blob/master/docs/development_notes.md#pull-requests) for more details.

haniotak committed:
adds integrated VPLS and LSP sync functionality
updates existing tests
minor changes to nso.sync property

zrecore commented on backend/src/main/java/net/es/oscars/sb/SouthboundPeriodicSyncer.java:
```diff
@@ -17,99 +21,191 @@
+                // this is a bit complicated logic
+                switch (syncState) {
+                    case DELETE -> {
+                        if (success) {
```
if "success" is true, why would the deployment state be marked as failed?

haniotak replied:
good catch, flipped the logic

zrecore commented on backend/src/main/java/net/es/oscars/sb/nso/NsoLspStateSyncer.java:
```diff
@@ -221,16 +227,25 @@ public boolean sync(String path, boolean dryRun) throws NsoStateSyncerException
                 for (NsoStateWrapper<NsoLSP> wrapper : toRedeploy) {
+                    NsoLSP lsp = wrapper.getInstance();
+                    if (!this.isOscarsManaged(lsp)) {
+                        log.error("attempted to redeploy unmanaged LSP " + lsp.getName() + "/" + lsp.getDevice());
```
either this should be a warning, or an exception should be thrown if truly an error.

zrecore commented on backend/src/main/java/net/es/oscars/sb/nso/NsoLspStateSyncer.java:
```diff
@@ -242,6 +257,11 @@ public boolean sync(String path, boolean dryRun) throws NsoStateSyncerException
                     List<NsoLSP> addList = new ArrayList<>();
                     NsoLSP lsp = wrapper.getInstance();
+                    if (!this.isOscarsManaged(lsp)) {
+                        log.error("attempted to add an unmanaged LSP " + lsp.getName() + "/" + lsp.getDevice());
```
same here... can we make this a proper error that must be handled, or switch to a warning?

haniotak replied:
changed those to log.warn

zrecore commented on backend/src/test/java/net/es/oscars/cuke/NsoStateManagerSteps.java:
```diff
@@ -54,29 +54,18 @@ public void the_state_manager_loads_VPLS_and_LSP_states() throws NsoStateManager
         knownInvalidVpls.add("SPaRC");
-        knownInvalidVpls.add("OSCARS-XYFK");
 
+        // make sure to skip unmanaged / invalid VPLS
```
Should this type of validation be part of the actual code base, instead of just the test?

haniotak replied:
it is in fact part of the codebase. as of now, VPLSs with vc-id outside the defined managed range (as per `nso.vc-id-range` property) are not added to the state. similar to what we were doing with LSPs that didn't follow the OSCARS name formatting.

this test ensures that those four unmanaged VPLSs that we know our NSO data includes, have not been added to the state. I renamed the test stuff a bit to match what it actually does.

haniotak committed:
fixes to address feedback

haniotak committed:
rename test for ignoring unmanaged vpls instances

haniotak committed:
modify vpls sync behavior;
VplsStateSyncer now also adds / redeploys associated LSPs together with their associated VPLS instance.

the reason is, when adding a VPLS instance to NSO that references an LSP instance, the LSP instance must be present (or added in the same commit)

haniotak committed:
fix flipped logic

haniotak committed:
stopped "description" from getting logged twice in quick succession

haniotak committed:
log lsps in nsoOscarsServices()
